metabrainz/invoices/send_invoices.py

```diff
@@ -0,0 +1,165 @@
+import datetime
+import io
+import time
+from decimal import Decimal
+
+from flask import current_app
+from intuitlib.enums import Scopes
+from intuitlib.client import AuthClient
+from quickbooks import QuickBooks
+from quickbooks.objects.customer import Customer
+from quickbooks.objects.invoice import Invoice, DeliveryInfo
+from quickbooks.objects.detailline import SalesItemLineDetail
+from quickbooks import exceptions
+from intuitlib.exceptions import AuthClientError
+
+from brainzutils import cache
+from metabrainz.mail import send_mail
+
+SEND_DELAY = 5
+MAIL_BODY = """To: %s %s (%s)
+
+Here is your next invoice! For information on how to pay us, please see https://metabrainz.org/payment
+
+Thanks!
+
+The MetaBrainz Foundation ( accounting@metabrainz.org ) 
+
+
+================================
+Invoice Summary:
+
+Invoice number: %s
+Invoice date: %s
+Due date: %s
+Invoice amount: %s%s
+================================
+
+Please see the attached PDF file for full details."""
+
+
+class QuickBooksInvoiceSender():
+    """ This class uses the QuickBooks API to sent invoices from our own mail ifrastructure. """
+
+    def __init__(self):
+        QuickBooks.enable_global()
+        self.auth_client = AuthClient(
+            client_id=current_app.config["QUICKBOOKS_CLIENT_ID"],
+            client_secret=current_app.config["QUICKBOOKS_CLIENT_SECRET"],
+            environment=current_app.config["QUICKBOOKS_SANDBOX"],
+            redirect_uri=current_app.config["QUICKBOOKS_CALLBACK_URL"]
+        )
+
+    def get_client(self):
+        """ Create the client object from the cached credientials and return it. """
+
+        refresh_token = cache.get("qb_refresh_token")
+        realm = cache.get("qb_realm")
+
+        if not refresh_token or not realm:
+            print("Could not fetch OAuth credentials from redis.")
+            print("Load https://metabrainz.org/admin/quickbooks/ to push the credentials to redis.")
+            return None
+
+        return QuickBooks(
+            auth_client=self.auth_client,
+            refresh_token=refresh_token,
+            company_id=realm
+        )
+
+    def mark_invoice_sent(self, client, invoice):
+        """ Given the client and invoice, mark this invoice as sent """
+
+        invoice.EmailStatus = "EmailSent"
+        if invoice.DeliveryInfo is None:
+            invoice.DeliveryInfo = DeliveryInfo()
+            invoice.DeliveryInfo.DeliveryType = "Email"
+
+        invoice.DeliveryInfo.DeliveryTime = datetime.datetime.strftime(datetime.datetime.now(), "%Y-%m-%dT%H:%M:%S")
+        try:
+            invoice.save(qb=client)
+            return True
+        except exceptions.ValidationException as err:
+            print(err.detail)
+            return False
+
+    def send_invoice(self, client, invoice, customer):
+        """ Given a client, invoice and its customer object, send the invoice to the customer. """
+
+        print("  sending invoice")
```

Comment:
Here and elsewhere in the file, instead of `print` can use `current_app.logger.info`